In [22]:
import numpy as np
import pandas as pd
from sklearn.linear_model import Perceptron
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.datasets.base import Bunch

В библиотеке scikit-learn линейные методы реализованы в пакете sklearn.linear_model. Мы будем работать с реализацией персептрона sklearn.linear_model.Perceptron. Как и у большинства моделей, обучение производится с помощью функции fit, построение прогнозов — с помощью функции predict.

В качестве метрики качества мы будем использовать долю верных ответов (accuracy). Для ее подсчета можно воспользоваться функцией sklearn.metrics.accuracy_score, первым аргументом которой является вектор правильных ответов, а вторым — вектор ответов алгоритма.

In [2]:
X = np.array([[1, 2], [3, 4], [5, 6]])
y = np.array([0, 1, 0])
clf = Perceptron()
clf.fit(X, y)
predictions = clf.predict(X)

Для стандартизации признаков удобно воспользоваться классом sklearn.preprocessing.StandardScaler. Функция fit_transform данного класса находит параметры нормализации (средние и дисперсии каждого признака) по выборке, и сразу же делает нормализацию выборки с использованием этих параметров. Функция transform делает нормализацию на основе уже найденных параметров.

In [6]:
scaler = StandardScaler()
X_train = np.array([[100.0, 2.0], [50.0, 4.0], [70.0, 6.0]])
X_test = np.array([[90.0, 1], [40.0, 3], [60.0, 4]])
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

#### 1
Загрузите обучающую и тестовую выборки из файлов perceptron-train.csv и perceptron-test.csv. Целевая переменная записана в первом столбце, признаки — во втором и третьем.

In [24]:
data = pd.read_csv('perceptron-train.csv', header=None)
train, test = Bunch(), Bunch()

train.data, train.target = data.loc[:, 1:], data.loc[:, 0]

data = pd.read_csv('perceptron-test.csv', header=None)
test.data, test.target = data.loc[:, 1:], data.loc[:, 0]

In [26]:
train.data.head()

1            2
0 -0.024626  1174.600238
1 -0.978058  1083.198803
2  0.314272 -1472.977609
3  0.179752   231.017267
4 -1.262544  -778.271726

In [27]:
train.target.head()

0   -1.0
1    1.0
2   -1.0
3   -1.0
4    1.0
Name: 0, dtype: float64

#### 2
Обучите персептрон со стандартными параметрами и random_state=241.

In [28]:
clf = Perceptron(random_state=241)
clf.fit(train.data, train.target)

Perceptron(alpha=0.0001, class_weight=None, eta0=1.0, fit_intercept=True,
      n_iter=5, n_jobs=1, penalty=None, random_state=241, shuffle=True,
      verbose=0, warm_start=False)

#### 3
Подсчитайте качество (долю правильно классифицированных объектов, accuracy) полученного классификатора на тестовой выборке.

In [33]:
accuracy = clf.score(test.data, test.target) # predicting
print(accuracy)

0.655


#### 4
Нормализуйте обучающую и тестовую выборку с помощью класса StandardScaler.

In [34]:
scaler = StandardScaler() # scaling
train_scaled, test_scaled = Bunch(), Bunch()
train_scaled.data = scaler.fit_transform(train.data)
test_scaled.data = scaler.transform(test.data)
train_scaled.target, test_scaled.target = train.target, test.target

#### 5
Обучите персептрон на новой выборке. Найдите долю правильных ответов на тестовой выборке.

In [35]:
clf.fit(train_scaled.data, train_scaled.target)

accuracy_scaled = clf.score(test_scaled.data, test_scaled.target)
print(accuracy_scaled)

0.845


#### 6
Найдите разность между качеством на тестовой выборке после нормализации и качеством до нее. Это число и будет ответом на задание.

In [37]:
difference = abs(accuracy_scaled - accuracy)
print(difference)

0.19


Нормализация данных дает значительный прирост в качестве — оно увеличивается на 19%.